File: solutions/java/src/main/java/com/epi/TreeDiameter.java
package com.epi;

import java.util.ArrayList;
import java.util.List;

public class TreeDiameter {
  // @include
  public static class TreeNode { List<Edge> edges = new ArrayList<>(); }

  private static class Edge {
    public TreeNode root;
    public Double length;

    public Edge(TreeNode root, Double length) {
      this.root = root;
      this.length = length;
    }
  }

  private static class HeightAndDiameter {
    public Double height;
    public Double diameter;

    public HeightAndDiameter(Double height, Double diameter) {
      this.height = height;
      this.diameter = diameter;
    }
  }

  public static double computeDiameter(TreeNode T) {
    return T != null ? computeHeightAndDiameter(T).diameter : 0.0;
  }

  private static HeightAndDiameter computeHeightAndDiameter(TreeNode r) {
    double diameter = Double.MIN_VALUE;
    double[] heights = {0.0, 0.0}; // Stores the max two heights.
    for (Edge e : r.edges) {
      HeightAndDiameter heightDiameter = computeHeightAndDiameter(e.root);
      if (heightDiameter.height + e.length > heights[0]) {
        heights[1] = heights[0];
        heights[0] = heightDiameter.height + e.length;
      } else if (heightDiameter.height + e.length > heights[1]) {
        heights[1] = heightDiameter.height + e.length;
      }
      diameter = Math.max(diameter, heightDiameter.diameter);
    }
    return new HeightAndDiameter(heights[0],
                                 Math.max(diameter, heights[0] + heights[1]));
  }
  // @exclude

  public static void main(String[] args) {
    TreeNode r = null;
    assert(0.0 == computeDiameter(r));
    r = new TreeNode();
    r.edges.add(new Edge(new TreeNode(), 10.0));
    r.edges.get(0).root.edges.add(new Edge(new TreeNode(), 50.0));
    r.edges.add(new Edge(new TreeNode(), 20.0));
    assert(80 == computeDiameter(r));
    System.out.println(computeDiameter(r));
    r.edges.get(0).root.edges.add(new Edge(new TreeNode(), 100.0));
    assert(150 == computeDiameter(r));
    System.out.println(computeDiameter(r));
  }
}

File: static/python/tree_diameter.py
import collections


# @include
class TreeNode:
    def __init__(self):
        self.edges = []


Edge = collections.namedtuple('Edge', ('root', 'length'))


def compute_diameter(T):
    HeightAndDiameter = collections.namedtuple('HeightAndDiameter',
                                               ('height', 'diameter'))

    def compute_height_and_diameter(r):
        diameter = float('-inf')
        heights = [0.0, 0.0]  # Stores the max two heights.
        for e in r.edges:
            h_d = compute_height_and_diameter(e.root)
            if h_d.height + e.length > heights[0]:
                heights = [h_d.height + e.length, heights[0]]
            elif h_d.height + e.length > heights[1]:
                heights[1] = h_d.height + e.length
            diameter = max(diameter, h_d.diameter)
        return HeightAndDiameter(heights[0],
                                 max(diameter, heights[0] + heights[1]))

    return compute_height_and_diameter(T).diameter if T else 0.0


# @exclude


def main():
    r = None
    assert 0.0 == compute_diameter(r)
    r = TreeNode()
    r.edges.append(Edge(TreeNode(), 10))
    r.edges[0].root.edges.append(Edge(TreeNode(), 50))
    r.edges.append(Edge(TreeNode(), 20))
    assert 80 == compute_diameter(r)
    print(compute_diameter(r))
    r.edges[0].root.edges.append(Edge(TreeNode(), 100))
    assert 150 == compute_diameter(r)
    print(compute_diameter(r))


if __name__ == '__main__':
    main()
